cores/esp8266/umm_malloc/umm_malloc_cfg.h

```diff
@@ -76,44 +56,242 @@ int _isr_safe_printf_P(const char *fmt, ...) __attribute__((format(printf, 1, 2)
 
-// umm_*alloc are not renamed to *alloc
+/* A couple of macros to make packing structures less compiler dependent */
 
-void *umm_malloc( size_t size );
-void *umm_calloc( size_t num, size_t size );
-void *umm_realloc( void *ptr, size_t size );
-#define umm_free    free
-#define umm_zalloc(s) umm_calloc(1,s)
+#define UMM_H_ATTPACKPRE
+#define UMM_H_ATTPACKSUF __attribute__((__packed__))
 
-void* malloc_loc (size_t s, const char* file, int line);
-void* calloc_loc (size_t n, size_t s, const char* file, int line);
-void* realloc_loc (void* p, size_t s, const char* file, int line);
+#define UMM_BEST_FIT
+#undef  UMM_FIRST_FIT
 
-// *alloc are macro calling *alloc_loc calling+checking umm_*alloc()
-// they are defined at the bottom of this file
+/*
+ * -D UMM_INFO :
+ *
+ * Enables a dup of the heap contents and a function to return the total
+ * heap size that is unallocated - note this is not the same as the largest
+ * unallocated block on the heap!
+ */
 
-/////////////////////////////////////////////////
-#else // !defined(ESP_DEBUG_OOM)
+#define UMM_INFO
+
+#ifdef UMM_INFO
+  typedef struct UMM_HEAP_INFO_t {
+    unsigned short int totalEntries;
+    unsigned short int usedEntries;
+    unsigned short int freeEntries;
+
+    unsigned short int totalBlocks;
+    unsigned short int usedBlocks;
+    unsigned short int freeBlocks;
+
+    unsigned short int maxFreeContiguousBlocks;
 
- // umm_*alloc are renamed to *alloc
- #define UMM_REDEFINE_MEM_FUNCTIONS
+    unsigned int freeSize2;
+  }
+  UMM_HEAP_INFO;
 
+  extern UMM_HEAP_INFO ummHeapInfo;
+
+  void ICACHE_FLASH_ATTR *umm_info( void *ptr, int force );
+  size_t ICACHE_FLASH_ATTR umm_free_heap_size_info( void );
+  size_t ICACHE_FLASH_ATTR umm_max_block_size( void );
+  size_t ICACHE_FLASH_ATTR umm_block_size( void );
+
+#else
 #endif
 
- #define UMM_BEST_FIT
+/*
+ * -D UMM_STATS :
+ *
+ * This option provides a lightweight alternative to using `umm_info` just for
+ * getting `umm_free_heap_size`.  With this option, a "free blocks" value is
+ * updated on each call to malloc/free/realloc. This option does not offer all
+ * the information that `umm_info` would have generated.
+ *
+ * This option is good for cases where the free heap is checked frequently. An
+ * example is when an app closely monitors free heap to detect memory leaks. In
+ * this case a single-core CPUs interrupt processing would have suffered the
+ * most.
+ */
+/*
+ */
+ #define UMM_STATS
 
-/* Start addresses and the size of the heap */
-extern char _heap_start[];
-#define UMM_MALLOC_CFG__HEAP_ADDR   ((uint32_t)&_heap_start)
-#define UMM_MALLOC_CFG__HEAP_SIZE   ((size_t)(0x3fffc000 - UMM_MALLOC_CFG__HEAP_ADDR))
+#ifdef DEBUG_ESP_PORT
+#define UMM_STATS_FULL
+#endif
 
-/* A couple of macros to make packing structures less compiler dependent */
+#if defined(UMM_STATS) || defined(UMM_STATS_FULL)
+
+typedef struct UMM_STATISTICS_t {
+  unsigned short int free_blocks;
+#ifdef UMM_STATS_FULL
```

Comment:
There were a few statistics I found useful for development; however, they all add to the IRAM usage. So I have isolated them out of the build via UMM_STATS_FULL. Are any of these useful to have around all the time? Are there some that are useful for debug enabled?

Reply:
@d-a-v implemented a metric for fragmentation which is. specific for us, i. e. it's not in upstream. I suggest making sure these stats are in line with that meteic, i. e. make sure the metric calculation code is unified, or if it isn't, that it's on purpose and for good reason. 
About your question, I think the oom count makes sense to have permanently. I'd use it as a health indicator, together with free heap, biggest free block size, and fragmentation metric. In my mind, if oom ever becomes > 0, it means something in the app needs checking at least, or fixing at worst. 
As for the rest, I don't see an immediate use, so maybe leave those for debug build.

Reply:
The fragmentation code was added to `umm_info` which walks the heap validating links and making a few computations along the way like finding the biggest free block. This can be time-intensive, i.e. exceeds 10us. 

Statistics under UMM_STATS are done with each malloc/realloc/free a small add/subtract max/min along the way.

Reply:
Let me ask in a different way: can the max free block and fragmentation metric be implemented as part of this stats? That wouldn't require walking the whole heap, and would therefore not be time intensive anymore, right? It would also unify the features and also reduce the number of places in the code with modifications.

Reply:
I think _max free block_ is the hard one, because as soon as you do an alloc that uses up some of the _max free block_ you need to search the free list for the next _max free block_. I don't see how to reliably make this one work w/o walking the whole heap.

_Fragmentation_ I think this could be done. It would require adjusting the square sum with every malloc/realloc/free. I think realloc will be the most painful one to handle. However, my concern here, is the added IRAM usage worth it? This seems more like a debug option than an always there type function.